File: day14/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	grid1 := map[Pos]struct{}{}
	grid2 := map[Pos]struct{}{}
	lowest := map[int]int{}
	floor := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		var points []Pos
		for _, i := range strings.Split(scanner.Text(), " -> ") {
			sx, sy, ok := strings.Cut(i, ",")
			if !ok {
				log.Fatalln("Invalid point")
			}
			x, err := strconv.Atoi(sx)
			if err != nil {
				log.Fatalln(err)
			}
			y, err := strconv.Atoi(sy)
			if err != nil {
				log.Fatalln(err)
			}
			points = append(points, Pos{y: y, x: x})
		}
		last := points[0]
		grid1[last] = struct{}{}
		grid2[last] = struct{}{}
		if v, ok := lowest[last.x]; !ok || last.y > v {
			lowest[last.x] = last.y
			if last.y > floor {
				floor = last.y
			}
		}
		for _, i := range points[1:] {
			delta := unitDelta(last, i)
			for last != i {
				last = addPos(last, delta)
				grid1[last] = struct{}{}
				grid2[last] = struct{}{}
				if v, ok := lowest[last.x]; !ok || last.y > v {
					lowest[last.x] = last.y
					if last.y > floor {
						floor = last.y
					}
				}
			}
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	floor += 2

	{
		count := 0
		for dropParticle1(Pos{y: 0, x: 500}, grid1, lowest) {
			count++
		}
		fmt.Println("Part 1:", count)
	}
	{
		count := 0
		for dropParticle2(Pos{y: 0, x: 500}, grid2, floor) {
			count++
		}
		fmt.Println("Part 2:", count)
	}
}

var (
	dirs = []Pos{
		{y: 1, x: 0},
		{y: 1, x: -1},
		{y: 1, x: 1},
	}
)

func dropParticle1(a Pos, grid map[Pos]struct{}, lowest map[int]int) bool {
	if _, ok := grid[a]; ok {
		return false
	}
outer:
	for {
		if v, ok := lowest[a.x]; !ok || a.y > v {
			return false
		}
		for _, i := range dirs {
			next := addPos(a, i)
			if _, ok := grid[next]; !ok {
				a = next
				continue outer
			}
		}
		grid[a] = struct{}{}
		return true
	}
}

func dropParticle2(a Pos, grid map[Pos]struct{}, floor int) bool {
	if _, ok := grid[a]; ok {
		return false
	}
outer:
	for {
		if a.y+1 < floor {
			for _, i := range dirs {
				next := addPos(a, i)
				if _, ok := grid[next]; !ok {
					a = next
					continue outer
				}
			}
		}
		grid[a] = struct{}{}
		return true
	}
}

type (
	Pos struct {
		y, x int
	}
)

func unitDir(a, b int) int {
	if a < b {
		return 1
	} else if a > b {
		return -1
	}
	return 0
}

func unitDelta(a, b Pos) Pos {
	return Pos{
		y: unitDir(a.y, b.y),
		x: unitDir(a.x, b.x),
	}
}

func addPos(a, b Pos) Pos {
	return Pos{
		y: a.y + b.y,
		x: a.x + b.x,
	}
}

File: day14/src/main.rs
use std::collections::hash_map::Entry;
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;
use std::ops::{Add, AddAssign};

const PUZZLEINPUT: &str = "input.txt";

type BoxError = Box<dyn std::error::Error>;
type BoxResult<T> = Result<T, BoxError>;

fn main() -> BoxResult<()> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut grid1 = HashSet::new();
    let mut grid2 = HashSet::new();
    let mut lowest = HashMap::new();
    let mut floor = 0;

    for line in reader.lines() {
        let mut first = true;
        let mut last = Pos::new(0, 0);
        for i in line?.split(" -> ") {
            let pos = if let Some((x, y)) = i.split_once(",") {
                Pos::new(x.parse::<i32>()?, y.parse::<i32>()?)
            } else {
                return Err("Invalid point".into());
            };
            if first {
                first = false;
                last = pos;
                grid1.insert(last);
                grid2.insert(last);
                match lowest.entry(last.x) {
                    Entry::Occupied(mut e) => {
                        let v = e.get_mut();
                        if last.y > *v {
                            *v = last.y;
                            if last.y > floor {
                                floor = last.y;
                            }
                        }
                    }
                    Entry::Vacant(e) => {
                        e.insert(last.y);
                        if last.y > floor {
                            floor = last.y;
                        }
                    }
                }
            } else {
                let delta = last.unit_delta(pos);
                while last != pos {
                    last += delta;
                    grid1.insert(last);
                    grid2.insert(last);
                    match lowest.entry(last.x) {
                        Entry::Occupied(mut e) => {
                            let v = e.get_mut();
                            if last.y > *v {
                                *v = last.y;
                                if last.y > floor {
                                    floor = last.y;
                                }
                            }
                        }
                        Entry::Vacant(e) => {
                            e.insert(last.y);
                            if last.y > floor {
                                floor = last.y;
                            }
                        }
                    }
                }
            }
        }
    }

    floor += 2;

    {
        let mut count = 0;
        while drop_particle1(Pos::new(500, 0), &mut grid1, &lowest) {
            count += 1;
        }
        println!("Part 1: {}", count);
    }

    {
        let mut count = 0;
        while drop_particle2(Pos::new(500, 0), &mut grid2, floor) {
            count += 1;
        }
        println!("Part 2: {}", count);
    }

    Ok(())
}

const DIRS: [Pos; 3] = [Pos::new(0, 1), Pos::new(-1, 1), Pos::new(1, 1)];

fn drop_particle1(mut a: Pos, grid: &mut HashSet<Pos>, lowest: &HashMap<i32, i32>) -> bool {
    if grid.contains(&a) {
        return false;
    }
    'outer: loop {
        match lowest.get(&a.x) {
            Some(&v) => {
                if a.y > v {
                    return false;
                }
            }
            None => return false,
        }
        for i in DIRS {
            let next = a + i;
            if !grid.contains(&next) {
                a = next;
                continue 'outer;
            }
        }
        grid.insert(a);
        return true;
    }
}

fn drop_particle2(mut a: Pos, grid: &mut HashSet<Pos>, floor: i32) -> bool {
    if grid.contains(&a) {
        return false;
    }
    'outer: loop {
        if a.y + 1 < floor {
            for i in DIRS {
                let next = a + i;
                if !grid.contains(&next) {
                    a = next;
                    continue 'outer;
                }
            }
        }
        grid.insert(a);
        return true;
    }
}

#[derive(Clone, Copy, Hash, PartialEq, Eq)]
struct Pos {
    y: i32,
    x: i32,
}

fn unit_dir(a: i32, b: i32) -> i32 {
    if a < b {
        1
    } else if a > b {
        -1
    } else {
        0
    }
}

impl Pos {
    const fn new(x: i32, y: i32) -> Self {
        Self { y, x }
    }

    fn unit_delta(&self, other: Self) -> Self {
        Self {
            y: unit_dir(self.y, other.y),
            x: unit_dir(self.x, other.x),
        }
    }
}

impl Add for Pos {
    type Output = Self;

    fn add(self, other: Self) -> Self {
        Self {
            y: self.y + other.y,
            x: self.x + other.x,
        }
    }
}

impl AddAssign for Pos {
    fn add_assign(&mut self, other: Self) {
        *self = Self {
            y: self.y + other.y,
            x: self.x + other.x,
        }
    }
}
